File: 270-closest-binary-search-tree-value/270-closest-binary-search-tree-value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    double diff = Double.MAX_VALUE;
    int res;
    public int closestValue(TreeNode root, double target) {
        inorder(root, target);
        return res;
    }
    public void inorder(TreeNode root, double target) {
        if(root == null) return;
        if(Math.abs(root.val - target) < diff) {
            diff = Math.abs(root.val - target);
            res = root.val;
        }
        if(root.val < target) inorder(root.right, target);
        else inorder(root.left, target);
    }
}

File: 270-closest-binary-search-tree-value/270-closest-binary-search-tree-value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        inorder_lst = []
        diff = float('inf')
        res = 0
        def inorder(root):
            nonlocal inorder_lst, res, diff
            if not root: return
            inorder(root.left)
            if abs(root.val - target) < diff:
                res = root.val
                diff = abs(root.val - target)
            inorder(root.right)
        inorder(root)
        return res